### ISSUE TITLE
Old Cut station only showing data to May 16 yet updated May 18

### ISSUE THREAD

**Issue by StuMackenzie** [MotusDev/Motus-TO-DO#970]
The Old Cut station is only reporting tags up to May 16 yet active tags remain and all other things seem OK up to May 18?

https://motus.org/data/receiverDeploymentDetections?id=5317

**Comment by jsayers-bsc** [comment]
There are no detections since the 16th, yet there doesn't appear to be anything wrong with the station. It's syncing, all antennas registering regular activity, GPS, raw data files on the server look fine. Just that the detections end a couple days ago. The obvious reason is that there no tags present, except that there is active tagging there still.

**Comment by StuMackenzie** [comment]
Missing tags - 355 371 356 306 314 312

**Comment by jsayers-bsc** [comment]
FYI Jeremy, I was at the site yesterday, confirming it's working and collecting and confirmed it's recognizing the tags that are present locally.

**Comment by Hussell** [comment]
Rerunning made it worse: hits from boot number 587 disappeared. File repo is intact, file permissions seem to be good, file sizes and contents look normal, but the tag finder is finding nothing. Metadata looks up-to-date. I'm not sure what changed.

**Comment by cjardine-bsc** [comment]
> 355 371 356 306 314 312

What project are these tags in? Or do you have motus tag_ids for them?

**Comment by StuMackenzie** [comment]
Project 20 Gugli.

I can activate other tags from another project and see if they’re being detected.

This problem is threat level orange!

Stu

____________________

**Comment by cjardine-bsc** [comment]
tag_ids for our reference: 

id	project_id	mfg_id	date_bin	type	codeset
52942	20	306	2021-2	ID	Lotek6M
52948	20	312	2021-2	ID	Lotek6M
52950	20	314	2021-2	ID	Lotek6M
52951	20	355	2021-2	ID	Lotek6M
52952	20	356	2021-2	ID	Lotek6M
52967	20	371	2021-2	ID	Lotek6M

@Hussell Note codeset. 

I currently have no idea what it could be. I'll look at the deprecated batch in Sandbox.

**Comment by StuMackenzie** [comment]
I’d include all 2021 tags from this set. Tags that were being detected stopped being detected and new tags aren’t being detected.

I hope beyond all hopes that this is an isolated issue.

____________________

**Comment by cjardine-bsc** [comment]
I think it is isolated.... 

I looked at the data coming in. We have ~900 SG taghits_daily records from May 14-16 and ~1000 from May 16-18. So tagfinder didn't just stop. 

Not that that helps anything except our immediate sanity.

**Comment by StuMackenzie** [comment]
The most important question I have now is, whether we should halt tagging, or will the data be ok once we sort out the issue.

____________________

**Comment by cjardine-bsc** [comment]
I don't have an answer to that excellent question yet. I'll look into what I can. But if Jeremy says the metadata is fine then this is something buried in the bowels of tagfinder.

**Comment by StuMackenzie** [comment]
I think number one priority is to figure out whether the problem is isolated to old cut or happening elsewhere. Can anyone think of scour for stations that have had data successfully run since May 16 shortly after midnight when this bug was introduced at Old Cur? Doing the same.

Best case scenario it’s restricted to old cut.

S

____________________

**Comment by cjardine-bsc** [comment]
Like I said, the amount of data coming in for the last two days has gone up from the previous two. So I don't think this is a wide spread issue. 

"I looked at the data coming in. We have ~900 SG taghits_daily records from May 14-16 and ~1000 from May 16-18. So tagfinder didn't just stop."

**Comment by StuMackenzie** [comment]
Another station at Powdermill stopped detecting tags May 12 - Project 35 - station PARC Banding ID7684 stopped recording tags May 12.

Tags were from project 416 and 246
____________________

**Comment by StuMackenzie** [comment]
Nix that. The Powdermill issue was the May 12 CTT bug. 

Onto our May 16 Motus bug.

____________________

**Comment by Hussell** [comment]
I can confirm that many stations are still seeing new detections, both by manual upload and through the automatic hourly uploads. PARC Banding (CTT-CB9A196520E9) hasn't had any new data uploaded since the 12th, so that's not the same problem.

The data files from Old Cut look normal, so I think if once this is figured out, nothing will have been lost.

**Comment by cjardine-bsc** [comment]
@hussell the problem is not the codeset. Tag_is 46839 is lotek4 and was detected in batchID 2382497 but not in your rerun batch 2398325

I agree the tag metadata looks fine. 

Unhelpful...

**Comment by denislepage** [comment]
Adding serial # for Old Cut, for reference: SG-5113BBBK2972

**Comment by jsayers-bsc** [comment]
@leberrigan I'm sending you raw data files for a [Project 48 receiver](https://motus.org/data/receiver/deployment/plot/SG%2D62E1RPI31AB3-2021%20NE%2DMotus%2C%20DB%20Farm.png?id=7736&n=100&firstTag=0). Can you run it through the unifile tagfinder? I'll also send the tag_database.

The tags should have been detected on the 17th. They were registered and a deployment created not long before that, however the detection data wasn't uploaded until the 19th. So the tags should have been detected. (They couldn't load the local tag db on the receiver, but did see the pulses of the tags in the pulses pane).

The tags that should have been detected:

tagID | label | BI | date_bin | model | freq | dt_start | &nbsp; | &nbsp;
-- | -- | -- | -- | -- | -- | -- | -- | --
54161 | 229 | 10.4997 | Feb-21 | NTQB2-5-1-M | 166.38 | 2021-05-17 - 2021-12-31
54160 | 228 | 10.4995 | Feb-21 | NTQB2-5-1-M | 166.38 | 2021-05-17 - 2021-12-31
54156 | 239 | 10.4993 | Feb-21 | NTQB2-3-2-M | 166.38 | 2021-05-17 - 2022-03-01
54155 | 238 | 10.5006 | Feb-21 | NTQB2-3-2-M | 166.38 | 2021-05-17 - 2022-03-01

**Comment by jsayers-bsc** [comment]
2 tags were present in the local tagfinder: 239 and 226 (not on the list). It's possible that they just got the Id's wrong when they told me since it was off-hand. 

But there are no tags on Motus.org for this receiver. @Hussell are you able to re-run this receiver SG-62E1RPI31AB3. Hopefully the tags show up.

**Comment by Hussell** [comment]
Bad news: someone spotted another receiver where the same thing happened, first detections not showing up, then a rerun causing previous detections to disappear. SG-2313BBBK3569

Rerunning SG-62E1RPI31AB3 also resulted in previous detections disappearing.

I'm pausing all further reruns until we fix this.

I'm going to try rolling back to beta9 for a day or two, in the hopes that it's a problem in the API providing metadata to the tag finder. That's one of the biggest things which changed recently, so best to rule it out.

**Comment by StuMackenzie** [comment]
Thanks Jeremy - if we can rollback to Beta 9, is there a way to quickly see if that 'fixes' it? fingers crossed

**Comment by StuMackenzie** [comment]
Which project is  SG-62E1RPI31AB3 associated with?

**Comment by Hussell** [comment]
SG-62E1RPI31AB3 is in project 48.

**Comment by Hussell** [comment]
> Thanks Jeremy - if we can rollback to Beta 9, is there a way to quickly see if that 'fixes' it? fingers crossed

If any of the batches where we've already observed loss of data are rerun and detections re-appear, then we'll have learned something important.

**Comment by StuMackenzie** [comment]
Do what you have to do! If we solve this quickly I'll sponsor you $10/bird on your birdathon.

**Comment by denislepage** [comment]
Just a thought, but could this be related to the recent changes regarding the more frequent loading of the metadata?

**Comment by denislepage** [comment]
I also just noticed that the 4 tags above are registered as Lotek6M. I would suggest trying to switch them to Lotek4 and see if they get picked up again, perhaps. I am not 100% sure we ever confirmed that the codeset was being properly handled??

**Comment by jsayers-bsc** [comment]
From using them as tests, we definitely confirmed the 6M tags are recognized and detected. Though they were also involved in some of the worst cases of aliasing I've seen (eg #619 - the blast at BSC as well as _14_ aliased tags resulting from just a single real tag). And now that I look, the new test tags not detected at BSC and a similar situation in Project 372 in addition to above are also all 6M. Could all be circumstantial though...

As an aside, wasn't Lotek delaying the launch of 6M until the SRX were all updated or something? I got the impression they weren't going to be issuing them for a while...

**Comment by Hussell** [comment]
Metadata refresh doesn't seem to be the cause. It's taking about 7 to 8 minutes, so I may reduce the frequency to once per 4 hours instead of hourly, but other than that it doesn't appear to affect anything. I've done several manual refreshes today, and tested reruns both during and between refreshes.

Codeset of the missing tags doesn't seem to be important, because rerunning is causing detections from non-6M tags to disappear. Perhaps worth looking into further.

I found and fixed a major bug involving the calculation of solar tag lifetimes. Unfortunately, it turned out not to be the cause of this issue, so you can skip the rest of this paragraph if you wish. I'm a bit disappointed this wasn't the cause, it looked really promising. The R code fitting a polynomial to the tag lifetimes reacted badly to having identical lifetimes for different duty cycles, resulting in wildly divergent end times. It was also producing some NA times, possibly due to exceeding the 2038 maximum Unix time. I've reduced the lifetime from 60 years to 10 years, and fixed the "singular gradient" error by giving the line of points a small slope. You can see the results on the [tag lifetime chart](https://sgdata.motus.org/public/motus_tag_timeline.html), e.g. tags 49365 (previously ended 2116) or 52210 (previously end time was NA). They're ending in 2036 now (2021 + 10 years * 1.5 buffer), so that gives us two years to make sure all tag finder code uses 64-bit integers for Unix time calculations.

**Comment by StuMackenzie** [comment]
Are we using Beta 9 yet, or is that still in progress?

____________________

**Comment by Hussell** [comment]
Yes, beta 9 has been on the beta server since ~10am. Doesn't seem to have had any effect, but I'll test again tomorrow.

**Comment by denislepage** [comment]
They haven’t released any new ID’s in 6M, but all tags in 4 match the for 521 tags in 6M, so they can be registered as either, and we have some registered as 6M in the database. There’s a couple different things that motusServer needs to do regardless, so that was a possible source of problem. But if those tags were previously detected as 6M, it’s unlikely to explain it.

**Comment by leberrigan** [comment]
I am still unable to download data from sgdata. I have ran the data the Josh provided for SG-62E1RPI31AB3 against NEMotus tag database and got 3 tags: 
- #97@166.38:39.7s, 
- #228@166.38:10.5s
- #239@166.38:10.5s

[SGdata 5.19 - OUTPUT.txt](https://github.com/MotusDev/Motus-TO-DO/files/6525424/SGdata.5.19.-.OUTPUT.txt)

**Comment by leberrigan** [comment]
According to the [find_tags Git repo](https://github.com/MotusWTS/find_tags) the only difference between the way the unifile version and motus versions of the software are looking for tags is that the unfile version ignores deployments and uses a static list of tags, whereas motus version does consider deployment periods. 

Considering I am able to detect tags in the unifile version of, but the motus version gives nothing, does this not indicate this issue is related to the tag event table?

**Comment by Hussell** [comment]
Not much progress over the weekend.
 * No changes were made to sgdata or the tag finder between May 6th and May 20th. (Recall we first noticed this on May 18th.)
 * Tags from all codesets are disappearing.
 * Reloading the tag codesets didn't change anything. (Try temporarily removing 6M?)
 * Rolling back to beta9 didn't change anything. (Also roll back changes to the beta/live database?)
 * After solar tags were fixed: there are no nulls in the tag events table and the minimum and maximum values look sane (2010, 2036).
 * Same for start and end times of tag deployments in sgdata's copy of the metadata.
 * Both receivers with lots of hits and receivers with few hits are affected. (So this probably isn't the result of hit IDs rolling over somewhere.)
 * Both connected and unconnected receivers are affected.
 * SensorGnome and CTT receivers are affected. Haven't yet found an affected Lotek receiver. (Not surprising, they use very different processes.)
 * Number of affected receivers is unclear. Need a way to find them in bulk.
 * Detections are still being found in new uploads. (Which ones? Haven't noticed a pattern yet, need to make a list.)

**Comment by Hussell** [comment]
Writing all that down shook something loose: I checked the max hit ID in the transfer tables and it's 4298508145, which is just over 2^32. I'm going to start scouring the tag finder looking for hit IDs stored as int32.

**Comment by StuMackenzie** [comment]
Good luck. go forth and prosper! If we aren't able to shake the fruit lose, we'll want to think about if/how we can easily summarize the problem for @joeybernard - if there is something for him to look for. Suppose your post above is probably as succinct and clear as we can get, but we'll want to provide a couple good examples.

**Comment by Hussell** [comment]
It doesn't really make sense, because the tag finder shouldn't be tracking the total number of hits across all receivers anywhere. Copying the data from receiver databases to the transfer tables is handled by sgdata (motusServer, written in R). But the tag finder is failing to insert hits into receiver databases, well before the transfer happens. So this might be unrelated, even though it looks really, really suspicious.

**Comment by Hussell** [comment]
* Restarting the whole server didn't fix the issue. (Had to check, just in case.)

**Comment by StuMackenzie** [comment]
Sigh...

____________________

**Comment by jsayers-bsc** [comment]
FYI Old Cut is syncing automatically again. 

The new deployment is here https://motus.org/data/receiverDeployent?id=7756 and the old one (active for several days) is here: https://motus.org/data/receiverDeployment?id=7743.

Both are up to date but neither have new detections (should be tags in the raw data for both)

**Comment by denislepage** [comment]
TBH, I don’t know if this is a good test because both units are attached to long time series of data. Do we have a new Rpi that has never been used that we could try to try to run a clean test?

If the test unit was already experiencing problems (do we know this?), attaching it to a different antenna likely won’t make any useful difference.

**Comment by Hussell** [comment]
* The new receiver at Old Cut (SG-5113BBBK2853) gave a bizarre result: 2 hits between the 21st and the 25th. Obviously far lower than it should be, but not strictly zero either.
 * Receivers in both Europe and the Americas are affected. Presumably in Australia too.

**Comment by leberrigan** [comment]
I just ran find_tags_unifile.exe on these three files:
```
> raw.files
[1] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T00-03-48.4370T-all.txt"
[2] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T01-03-51.1270T-all.txt"
[3] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T02-03-51.4520T-all.txt"
[4] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T03-03-53.3050T-all.txt"
[5] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T04-03-53.9150T-all.txt"
[6] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T05-03-54.3810T-all.txt"
[7] "Old cut/SG1-5113BBBK2853-000404-2021-05-25T06-03-55.5740T-all.txt"
```
Using the most recent tagdb from project #20 and got the following results:
![tagplot_oldcut](https://user-images.githubusercontent.com/35931796/119822473-b2a42700-bec1-11eb-84b5-176998ed741c.png)

**Comment by leberrigan** [comment]
If we're getting this error regardless of what receiver is there, it must be a tag metadata issue. We should first address Denis point and test this again with a new receiver, not just a different one. You can just create a test deployment for that receiver; it doesn't matter what the deploy name or ID is when processing data. 

If we are still not seeing tags detected, there should be a tag test. If you have test tags, bring them to the Old Cut station to see if they're picked up. If you can get your hands on any of the Gugli tags that are not being detected but should have been, bring that to a station that is currently working (BSC_HQ?).

**Comment by denislepage** [comment]
We didn’t discuss this today, but do we have a plan to test with a new RPi?

**Comment by jsayers-bsc** [comment]
I was going to ensure one is hooked up today at HQ along with some active tags. These are new and I don't believe they have uploaded any data yet (though they may have been connected to the server earlier as a test)

**Comment by StuMackenzie** [comment]
Josh and I are working on this. Hopefully soon

____________________

**Comment by jsayers-bsc** [comment]
> 
> We didn’t discuss this today, but do we have a plan to test with a new RPi?

Apparently the only completely new Pi has been plugged in and likely syncing over the last couple weeks. I say likely because I've having trouble connecting to it (I was testing it with multiple 166 and 434 antennas via a hub and it's possible that part hasn't been working well)

**Comment by jsayers-bsc** [comment]
There are 3 active test tags at BSC and two active receiver deployments.

- https://motus.org/data/receiverDeployment?id=7695. Main BSC HQ. station. Swapped a SS a couple weeks ago and it's connected via internet. It's never shown a 166 tag detection though there have been some here (not all the time until now, but certainly something since the 16th)
- https://motus.org/data/receiverDeployment?id=7761. Just set up to replace the one mentioned above. Just has 3 FCD and is beside the 3 tests. Connected to internet but hasn't had time to send data yet.

**Comment by joeybernard** [comment]
I've looked into the databases on sgdata. There are 8 SQLite databases in /sgm_local, and the biggest file is just a bit over 4GB in size. The largest number of rows in this DB is 6,483,436. So, while things are getting large they aren't hitting limits yet. I guess the next step is to add logging to find_tags_motus so we get some output. Jeremy, I'll let you know once it is ready to put in place.

**Comment by Hussell** [comment]
Running the enhanced logging branch of the tag finder produced segmentation faults, crashing the program before any log messages were outputted. Logs are at /sgm_local/logs/find_tags.log (currently 3 lines from a test last night).

https://sgdata.motus.org/status?jobID=7022675

**Comment by denislepage** [comment]
Not very familiar with C++, but according to this thread, a segfault is often caused by a Null Pointer.
https://stackoverflow.com/questions/2346806/what-is-a-segmentation-fault

There are other ways, but given that we’ve been operating this code for years now without that problem, it’s much more likely something related to the metadata than the data or the code itself, is my guess.

Well, that feels at least like a bit of progress. At least we’re getting an error.

Joey, are you able to find time this week to add more logging prior or around this around?

**Comment by joeybernard** [comment]
OK. Why don't you go back to using the original version for now. I have the data file copied into my home directory and I will try getting it to run to completion. I'm hoping to get some tonight to do it.

**Comment by joeybernard** [comment]
OK, so this is what I have figured out so far. When I tried the command from the rerun that Jeremy sent me, I was getting errors about malformed lines. I was able to track it down to the file 

/sgm/file_repo/Binbrook-5113BBBK2972-000588-2021-05-16T22-28-15.8490T-all.txt

In this file, there is a line that has the start as "p2G" rather than "p2". The problem here is that the source code file "SG_Record.cpp" expects that the port (after the record type) is a single character. In almost all cases, this is a single number. In this particular case, the port is "2G". This means that the rest of the processing for the line is messed up, and I think this might be the source of the issue. What do you guys think?

**Comment by StuMackenzie** [comment]
Could be, but we'd have to show that it's repeating in all of the other affected receivers. Then of course the question is where is it coming from? We can prioritize/provide a list of other receivers to dig into if you think that's a good next step - some are listed above, but it's a bit garbled.

**Comment by denislepage** [comment]
Possible. But odd. Why have those suddenly become more prevalent, and what does the G extension mean? @dfloer could you investigate if p2G has a specific meaning? The options we have is to ignore those lines or perform a correction to force a numeric value (either options I imagine are doable, @joeybernard?). We may also need to figure out what other patterns we need to look for. 

My preference would be to have a way to catch errors like this on individual line and provide some kind of counter at the end (e.g. XXXX lines could not be read because they were invalidly formatted). If we know some of the exceptions to look for, we should try to accommodate them the best we can.

Perhaps the issue is rare and the receiver will be processed up to the point where the p2G entry happens. What is odd is that re-running older data also did not work, but perhaps it does to the end of the batch before starting tag recognition.

@Hussell , I meant to ask this, but are you able to re-run a batch, but exclude the recent files where the problem started happening? This would likely conform with (though not confirm) Joey's finding.

@joeybernard  could we try creating a patch again that would allow by-passing the error until we can sort out the meaning of G.

**Comment by denislepage** [comment]
The other thing is whether we'd ever have more than 9 ports, and whether this would also cause similar issues. E.g. p12. SensorGnome doesn't support that many, but I think you could in theory have a USB hub to extend it, yes?

**Comment by StuMackenzie** [comment]
Could but we typically stop at 8. Most are 4 or less.

____________________

**Comment by denislepage** [comment]
To clarify, the full path of the file above is as follow:

/sgm/file_repo/SG-5113BBBK2972/2021-05-16/Binbrook-5113BBBK2972-000588-2021-05-16T22-28-15.8490T-all.txt

Here's the bit that was corrupted. The data associated with p2G is actually GPS data, so the p2 is incorrectly inserted there. The following line should be a pulse, but it is missing the port number.

p2,1621205629.2217,-7.603,-67.43,-75.25
p2,1621205629.2235,11.66,-70.52,-78.25
**p2G,1621205628,42.582821667,-80.398133333,183.1
,1621205629.2242,3.924,-43.52,-63.94**
p2,1621205629.2265,6.613,-66.84,-75.56
p2,1621205629.3047,3.924,-43.62,-76.46

I have edited the file to remove the 2 offending lines and saved it back on the drive. Can we try again to run that receiver??? I can't believe I'm going to have to wait until tomorrow (the suspense is killing me!!!)

**Comment by denislepage** [comment]
This is likely due to some concurrent events both trying to write to the file at the same time, which would happen only rarely, but with hundred of billions of pulses, it's bound to happen occasionally. 

This could advocate for NOT trying to correct those lines, and simply skipping over them. This would potentially add a bit of overhead if we have to validate every line. The other option would be to do the line parsing operations within a try/catch block.

How feasible is that @joeybernard ? 

I had to take similar steps in reading CTT files as they also show odd patterns once in a blue moon, and this was also stopping the processing of individual batches.

**Comment by denislepage** [comment]
https://www.youtube.com/watch?v=oIVSeWBk4qM

**Comment by denislepage** [comment]
Given that the processing stopped on May 16, which is when the corrupted file turned up, it would certainly make a WHOLE lot of sense. If find_tags is not able to deal with a single corrupted data line, then this is likely the main problem we're seeing elsewhere. 

Next step: make find_tag more robust to invalid data, and anticipate the unexpected with proper try/catch at a granular level and proper associated logging to help alert us of these issues.

If we know when other receivers have stopped working, it should be easy enough to look for a corrupted file around that time. Or, just patch the software and run the batches again and see. I would at least try to examine a couple more cases to see if the problem is similar or different.

**Comment by leberrigan** [comment]
First thing that comes to mind with "p2G" is exactly what Denis said, that they're trying to write the same line concurrently.  I am very surprised that tag finder is _this_ fragile.

Now I am wondering why we've only noticed this now.. Very odd that it would have occurred for the first time only now and on multiple receivers with no changes made.

In [`find_tags/src/SG_Record.cpp`](https://github.com/MotusWTS/find_tags/blob/2d6399fb0cc241604208dd47150352f28f3730cf/src/SG_Record.cpp), we can see (line 16) the case where 'p' is identified as the first character in the line:

```
SG_Record::from_buf(char * buf) {
  // assume invalid record
  type = BAD;
  switch (buf[0]) {
  case 'p':
    /* a pulse record line like:
       p1,14332651182.1235,3.234,-55.44,-77.33
       port    ts          dfreq  sig    noise
    */
    if (5 == sscanf(buf+1, "%hd,%lf,%f,%f,%f", &port, &ts, &v.dfreq, &v.sig, &v.noise)) {
      type = PULSE;
    }
    break;

...
```

According to the [documentation for `sscanf`](https://www.cplusplus.com/reference/cstdio/sscanf/), this function will return either an error (if input error) or number of matches (if some aren't matched). 
Based on this, `sscanf` should return **4** for the first line

> "p2G,1621205628,42.582821667,-80.398133333,183.1"

This is because because the first item in the list is expected to be a `short int` and it is actually "2G", so the `type` variable remains as `BAD`. 

The second line won't have a match in any of the cases so it gets defaulted and also remains as `type = BAD`
```
  default:
    break;
  };
```

**Comment by joeybernard** [comment]
Also, I found this late last night after several hours. With the light of a new day, I am no longer so sure that this has anything to do with this issue. I am thinking now that this may just be a coincidence.

**Comment by leberrigan** [comment]
It's good to make sure, though. It seems like there was an [edit to `SG_Record.cpp`](https://github.com/MotusWTS/find_tags/commit/5b1b67cf80f4fd0b203c6fde994379723355b842#diff-266c125fd455117ed488273f9219637f5aa01c39a1575c64eb07ec92ec056140) last August so perhaps this has only been a problem for a short time and due to the rarity of malformed lines in this manner it hasn't been picked up until now.

These changes allow for 3 or 4 matches in the GPS fix variables to allow for missing altitudes, but does not check to makes sure the altitude is _actually_ missing, instead of it being a malformed line. I wonder if a malformed GPS line could produce errors in this way.

Perhaps we could revert these changes temporarily to see if it fixes anything?

**Comment by joeybernard** [comment]
I can give this a try this evening with the test config I have set up.

**Comment by Hussell** [comment]
So far as I can see, these two lines are skipped. The first triggers case 'p', then fails to match, and the second goes through to the default case. Case 'G', where the most recent changes are, is not triggered.

When I looked at the file in /sgm/file_repo/SG-5113BBBK2972/2021-05-16/, it was changed today, but the p2G line was still there. I've made a backup and replaced it with a version with those two lines fixed. Rerunning with a rescan of the file_repo files will take a while, but I strongly suspect this is not the source of our trouble. The tag finder has correctly ignored far more corrupt lines in the past, e.g. ones which included sections of random binary gibberish.

**Comment by joeybernard** [comment]
OK, I tried this with no luck. I have added a bunch more debugging output and I think we need "real" output from production.

Jeremy, I put a binary called "find_tags_motus_enhanced_logging" in "/sgm_local/bin". When you get a chance, could you rename it to "find_tags_motus" and dp a rerun that you know had issues before. Hopefully we get something useful in the log file after that. Once the rerun is done, you can copy back in the original binary until we have time to review the log file.

**Comment by Hussell** [comment]
I ran one boot number from Old Cut, and that produced a 3GB log file. Should I look for a smaller batch to rerun?

**Comment by Hussell** [comment]
Looks like I goofed and overwrote the only copy of the executable with enhanced logging when I was putting the original back in place, so I'll have to wait until tomorrow to run a smaller batch.

**Comment by denislepage** [comment]
Woomp-woomp.

**Comment by joeybernard** [comment]
Well, it looks like find_tags_motus is not crashing. The logs show that it gets all the way to the end.

I have added even more detailed logging and put a freshly compiled binary in /sgm_local/bin.  Jeremy, could you do another rerun whenever you get a chance? Hopefully I am dumping enough detail now to help figure out what is happening.

**Comment by Hussell** [comment]
Done.

**Comment by Hussell** [comment]
I've created a log for SG-5113BBBK2853 boot number 404. That's the receiver substituted into Old Cut on May 21st, and the boot number which generates 2 hits instead of 0 (full failure) or many (expected result). Some of lines are different than the other logs I've looked at, so there may be a clue there to what's going on.

**Comment by Hussell** [comment]
Alright, there are lines about tags being "activated" in all the logs, but lines about tags being "added" only appear in logs for runs which produce detections. I was able to verify that the tag detected in SG-5113BBBK2853 boot number 404 was one of the tags which was added, and that a lot of tags active during May are never added. It would be helpful to find out from Joey where in the code he inserted the log message "Tag addedd", since it hasn't been added to the master or enhanced_logging branches.

This looks like a promising lead.

**Comment by joeybernard** [comment]
OK, here is where the output is coming from:

- Activating or deactivating a tag - from the process_event function in TagForay.cpp.
- Tag added or deleted - from the tag_added or tag_deleted functions in Tag_Finder.cpp, this is called from Tag_Foray.cpp when processing an E_ACTIVATE or E_DEACTIVATE event.

Looking at the output, there are lot of tags that get deleted or deactivated when either they appear as expired or when the offset frequency goes out of bounds.

If people can think of more information that will help, let me know and I can add more debugging in the appropraite places.

**Comment by denislepage** [comment]
That’d suggest one of those 2 things (dates or frequency offsets) are getting messed up, or perhaps the BI period, etc.

Adding some logging that track those values would probably be helpful, and the boundaries used for each candidate. The latter could potentially be logged once at the start of a batch, and the individual values against which they are compared logged when a tag_deleted event happens. If we know specifically what tag we are expecting to see based on prior results, we could also examine a few of these more carefully (e.g. log any occurrences matching these tags)

The problem is either in one of those comparison to assess the rules about tag_deleted, or the values used are messed up before they get there, somehow.

I doubt this will help, but I examined a little bit the metadata that was saved to the git repo earlier today: https://github.com/jbrzusto/motus-metadata-history

We have 31,047 Lotek tags and 31,919 tag deployments. The longest tag model name we have is "NTQB2-4-2S-M" (12 chars). For a detection happening today, there would be 3692 eligible tags to identify (2958 if you limit this to 166.38 tags matching Old Cut antennae)

If we had a way to track which tags are eligible during processing (or at least a count that we can compare to the numbers above, etc.), that might give us a clue.

Finally, I doubt this is relevant, because these were registered a while back (2015 and 2018), but I found 6 tags that have deployment info, but are not in the tags list. 8438, 8439, 31896, 31897, 31898, 31899. This is odd. The database currently has 30,747 Lotek tags (400 less than the metadata used by SGdata). and 29,506 deployments (2,400 less than sgdata). This suggests that tags that are removed in Motus at not taken out from sgdata’s metadata, but I think we already knew this. The 6 above are no longer in the database (but we have them in our backup archive, not visible to sgdata) As long as these 6 tags are not causing any problems, we can leave this one alone for now.

Ideally, we’d have some level of control over the granularity of the logging in the future. The default could be only to report errors, and there could be 1 or 2 extra levels (e.g. warnings and info) that we could turn on manually when we need. Something like that. Probably something we can deal with once this problem is resolved. Alternating branches is fine for now.

**Comment by StuMackenzie** [comment]
For documentation: Another SG-F790RPI3E8DD also showing no detections; Project 113.

When the time is right we'll have a short list of projects to prioritize reprocessing.

**Comment by joeybernard** [comment]
OK, I added more logging around tags. Jeremy, could you do a rerun or two when you get a chance tomorrow? Hopefully this will do it?

As for what we do after this emergency is done - I will be cleaning up the debug code I have added and also add a new command line option. This will control the debug level for any particular run by simply changing the options for a run. This way, we'll be able to get similar output the next time an issue comes up.

**Comment by Hussell** [comment]
Holy shit, I figured it out!

**Comment by StuMackenzie** [comment]
Not sure I've ever heard, or seen, you swear.... patiently waiting emoji...

**Comment by Hussell** [comment]
I still don't understand exactly what's going wrong inside the tag finder, but the cause is in the metadata: 200 tags, first registered on the 14th of May, with nomFreq = 166.376. When I change these to something else in the metadata DB on sgdata (either 166.38 or something wildly different like 999), then rerun, the missing detections show up. I haven't checked them all, but it looks like these tags are the ones mentioned above as getting "added" where no other tags were.

In the code it looks as though tags and/or tag finders are being indexed by nominal frequency, so the bug probably involves an assumption that all nominal tag frequencies within the receiver's frequency window are identical.

**Comment by Hussell** [comment]
Can we set up a quick Zoom call?

**Comment by StuMackenzie** [comment]
I'll check Deni's schedule and 
send around an invite

**Comment by StuMackenzie** [comment]
Well, calendars can be wrong so we're all free now, or 300. I'll logon now regardless.

Stuart Mackenzie is inviting you to a scheduled Zoom meeting.

Topic: Stuart Mackenzie's Personal Meeting Room

Join Zoom Meeting
https://birdscanada-org.zoom.us/j/6772244522?pwd=MWxqOE52d1loZ1NYZGpMUGxzUUlGdz09

Meeting ID: 677 224 4522
Passcode: 182516
One tap mobile
+16475580588,,6772244522#,,,,*182516# Canada
+17789072071,,6772244522#,,,,*182516# Canada

Dial by your location
        +1 647 558 0588 Canada
        +1 778 907 2071 Canada
        +1 204 272 7920 Canada
        +1 438 809 7799 Canada
        +1 587 328 1099 Canada
        +1 647 374 4685 Canada
        +1 669 900 6833 US (San Jose)
        +1 929 205 6099 US (New York)
        +1 253 215 8782 US (Tacoma)
        +1 301 715 8592 US (Washington DC)
        +1 312 626 6799 US (Chicago)
        +1 346 248 7799 US (Houston)
Meeting ID: 677 224 4522
Passcode: 182516
Find your local number: https://birdscanada-org.zoom.us/u/kclGoTXqJe

**Comment by denislepage** [comment]
About to have lunch, but I’ll join shortly.

**Comment by StuMackenzie** [comment]
*Priority Projects to re-run*

- [ ] Project 1

_- [ ] Project 113 - separate issue?_

- [ ] Project 226

- [ ] Project 213

- [ ] Project 48 - _big receiver project with a lot of 'live'receivers_

@jsayers-bsc and @leberrigan please add others you may have dealt with.

**Comment by leberrigan** [comment]
Project 113: 
- Deploy ID #6163 (SG-50D6RPI38864)
  - No detections since 21 Apr 2021, but nearby stations (5-10kmkm) have detected many tags up to the end of May.
  - Lots of tags are picked up when running raw data thru find_tags_unifile, according to researchers.

**Comment by Hussell** [comment]
The SG-50D6RPI38864 boot number I reran during the call was the wrong one (9 instead of 10), after rerunning the whole thing there are detections: https://motus.org/data/receiver/deployment/plot?id=6163

**Comment by leberrigan** [comment]
Thanks Jeremy - any idea what happened there? It's probably worth re-running stations in this project with recent data as well.

**Comment by Hussell** [comment]
Probably the same issue they had before, where until the first rerun no detections would be found, made more obvious by the 3-week pause of reruns. Digging into that again will have to wait until we're done with this one and caught up with reruns.

**Comment by jsayers-bsc** [comment]
I was in the field today and missed the discussion, but to clarify Jeremy you think this may be the same as the "European Receiver" issue?

**Comment by Hussell** [comment]
No, definitely not. In fact, I think European receivers were not affected, despite what I thought earlier.

**Comment by Hussell** [comment]
Reprocessing is done. There are a handful of batches from between 2021-05-14 and 2021-06-09 which for one reason or another haven't yet been deprecated, but they've all been replaced by newer versions. I'll clean those up shortly.

The remaining tasks: fix the root bug, and integrate logging code from @joeybernard.

**Comment by Hussell** [comment]
Looks like someone changed a bunch of tag frequencies to 166.376 yesterday. 361 tags from projects 57, 188, 199, 232, 269, and 270. About evenly split between tags from 2017 and 2021. The user ID of the last editor isn't set in most cases, so it might not have been done through the website?

Does anyone have any further clues?

**Comment by Hussell** [comment]
It looks as though almost all our tag metadata was updated at precisely the same time yesterday: 32,759 tags have last_edited_dt = '2021-07-14 16:27:30.983', vs 6,414 with other times. @cjardine-bsc, do you have any idea why this might be?

**Comment by denislepage** [comment]
I am sure Catherine is involved, but what I am not seeing in this thread is whether this is causing any issues?

**Comment by Hussell** [comment]
Yes it's causing issues. Once the frequencies are reset to 166.38 I'm going to have to rerun everything since yesterday. (Not very much, thankfully.)

**Comment by cjardine-bsc** [comment]
Metadata should be back in order. Sorry for the scare :-)

**Comment by StuMackenzie** [comment]
What's the final cleanup step here, or can we close?

**Comment by cjardine-bsc** [comment]
Final cleanup could include preventing any more bad metadata by moving to lotek registration process. 

True fix would be a change to tagfinder so it doesn't implode when given bad metadata.
